src/generic/Matrix.jl

```diff
@@ -5260,6 +5248,40 @@ function identity_matrix(M::MatElem{T}, n::Int) where T <: RingElement
    return identity_matrix(base_ring(M), n, n)
 end
 
+################################################################################
+#
+#   Diagonal matrix
+#
+################################################################################
+
+@doc Markdown.doc"""
+    diagonal_matrix(x::RingElement, m::Int, [n::Int])
+> Return the $m \times n$ matrix over $R$ with `x` along the main diagonal and
+> zeroes elsewhere. If `n` is not specified, it defaults to `m`.
+
+# Examples
+
+```jldoctest
+julia> diagonal_matrix(ZZ(2), 2, 3)
+[2  0  0]
+[0  2  0]
+
+julia> diagonal_matrix(QQ(-1), 3)
+[-1//1   0//1   0//1]
+[ 0//1  -1//1   0//1]
+[ 0//1   0//1  -1//1]
+```
+"""
+function diagonal_matrix(x::RingElement, m::Int, n::Int)
+   z = zero_matrix(parent(x), m, n)
+   for i in 1:min(m, n)
+      z[i, i] = deepcopy(x)
```

Comment:
Is the deepcopy required? As per Julia, we do not assume that the entries of our (generic) matrices need to be unique, and all other matrix types make copies when inserting data anyway (e.g. Flint matrices).

So I *think* it's ok to drop the deepcopy. If someone wants to modify an entry of a matrix, they have to replace it. If they want to mutate the actual element there, they have to own it.

Admittedly it is not completely clear how they can create a matrix whose elements they own exclusively, without creating all the entries themselves, but still...

Reply:
In fact, by rule 4 here:

https://github.com/Nemocas/Nemo.jl/issues/278

It definitely is not required to deepcopy the value. The same is true of zero_matrix, actually, just in case it is done there.

Reply:
Hmm, now I am myself confused. In Matrix.jl, e.g. in hnf_cohen!, it clearly mutates existing entries of a matrix, which seems to violate rule 4. Probably this happens often in the delayed reduction code throughout Matrix.jl.

We had better fix this. However, I first have to understand the correct fix.

Reply:
Nope, this was just a clear bug in the code in Matrix.jl. The rules are intact. I will shortly issue a PR to fix this. It was clear someone had previously tried to fix it, but failed to actually finish it.

Reply:
So yes, the deepcopy should not be necessary here (or in zero_matrix, if you have one).

Reply:
Oh I'm very happy to hear that, I haven't internalized yet the rules for when `deepcopy` is necessary. I indeed did like this by taking `zero_matrix` as an example, and also because the implementation of `identity_matrix`, which is now delegated to `diagonal_matrix` here, was doing this too. Ok I will update, and maybe make another PR to avoid unnecessary copies.